In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
df = pd.read_csv("fertilizer_recommendation.csv")

print("Dataset loaded successfully.")
print("Shape:", df.shape)

Dataset loaded successfully.
Shape: (10000, 20)


In [3]:
display(df.head())

,Soil_Type,Soil_pH,Soil_Moisture,Organic_Carbon,Electrical_Conductivity,Nitrogen_Level,Phosphorus_Level,Potassium_Level,Temperature,Humidity,Rainfall,Crop_Type,Crop_Growth_Stage,Season,Irrigation_Type,Previous_Crop,Region,Fertilizer_Used_Last_Season,Yield_Last_Season,Recommended_Fertilizer
0,Clay,6.07,34.98,0.32,1.87,61,44,84,19.84,83.31,1693.22,Cotton,Harvest,Kharif,Canal,Wheat,South,297.15,1.19,MOP
1,Silt,6.39,47.34,0.28,0.21,59,56,18,24.40,46.27,1030.21,Maize,Vegetative,Kharif,Sprinkler,Potato,Central,77.17,4.40,Urea
2,Sandy,7.92,38.13,0.99,1.88,43,21,119,24.82,71.86,1166.16,Cotton,Flowering,Kharif,Rainfed,Tomato,South,128.93,7.21,Urea
3,Clay,5.86,14.17,1.46,0.36,88,46,34,27.87,53.23,2881.83,Wheat,Flowering,Zaid,Sprinkler,Potato,West,233.96,1.85,MOP
4,Clay,7.98,19.28,0.85,2.16,104,53,98,24.17,51.87,714.84,Potato,Sowing,Kharif,Rainfed,Maize,East,214.39,7.36,Zinc Sulphate


In [4]:
print(df.columns.tolist())

['Soil_Type', 'Soil_pH', 'Soil_Moisture', 'Organic_Carbon', 'Electrical_Conductivity', 'Nitrogen_Level', 'Phosphorus_Level', 'Potassium_Level', 'Temperature', 'Humidity', 'Rainfall', 'Crop_Type', 'Crop_Growth_Stage', 'Season', 'Irrigation_Type', 'Previous_Crop', 'Region', 'Fertilizer_Used_Last_Season', 'Yield_Last_Season', 'Recommended_Fertilizer']


In [5]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:")
print(df.duplicated().sum())

Rows: 10000
Columns: 20

Missing values:
Soil_Type                      0
Soil_pH                        0
Soil_Moisture                  0
Organic_Carbon                 0
Electrical_Conductivity        0
Nitrogen_Level                 0
Phosphorus_Level               0
Potassium_Level                0
Temperature                    0
Humidity                       0
Rainfall                       0
Crop_Type                      0
Crop_Growth_Stage              0
Season                         0
Irrigation_Type                0
Previous_Crop                  0
Region                         0
Fertilizer_Used_Last_Season    0
Yield_Last_Season              0
Recommended_Fertilizer         0
dtype: int64

Duplicate rows:
0


In [6]:
print(df["Crop_Type"].value_counts())

Crop_Type
Wheat        1467
Tomato       1451
Rice         1429
Sugarcane    1428
Potato       1425
Maize        1424
Cotton       1376
Name: count, dtype: int64


In [7]:
fertilizer_df = df[
    df["Crop_Type"].isin(["Rice", "Sugarcane"])
].copy()

print("Filtered dataset shape:", fertilizer_df.shape)

print("\nCrop distribution:")
print(fertilizer_df["Crop_Type"].value_counts())

Filtered dataset shape: (2857, 20)

Crop distribution:
Crop_Type
Rice         1429
Sugarcane    1428
Name: count, dtype: int64


In [8]:
print(
    fertilizer_df["Recommended_Fertilizer"]
    .value_counts()
)

Recommended_Fertilizer
Urea             912
DAP              838
MOP              394
Compost          281
Zinc Sulphate    203
NPK              182
SSP               47
Name: count, dtype: int64


In [9]:
print(
    fertilizer_df[
        "Recommended_Fertilizer"
    ].value_counts(normalize=True).round(3)
)

Recommended_Fertilizer
Urea             0.319
DAP              0.293
MOP              0.138
Compost          0.098
Zinc Sulphate    0.071
NPK              0.064
SSP              0.016
Name: proportion, dtype: float64


In [10]:
crop_fertilizer_table = pd.crosstab(
    fertilizer_df["Crop_Type"],
    fertilizer_df["Recommended_Fertilizer"]
)

display(crop_fertilizer_table)

Recommended_Fertilizer,Compost,DAP,MOP,NPK,SSP,Urea,Zinc Sulphate
Crop_Type,,,,,,,
Rice,145,414,193,103,17,458,99
Sugarcane,136,424,201,79,30,454,104


In [11]:
numeric_features = [
    "Soil_pH",
    "Soil_Moisture",
    "Organic_Carbon",
    "Electrical_Conductivity",
    "Nitrogen_Level",
    "Phosphorus_Level",
    "Potassium_Level",
    "Temperature",
    "Humidity",
    "Rainfall",
    "Fertilizer_Used_Last_Season",
    "Yield_Last_Season"
]

display(
    fertilizer_df[numeric_features]
    .describe()
    .T
)

,count,mean,std,min,25%,50%,75%,max
Soil_pH,2857.0,6.509790,1.152211,4.50,5.51,6.49,7.55,8.50
Soil_Moisture,2857.0,35.344123,14.510438,10.00,22.83,35.71,47.34,60.00
Organic_Carbon,2857.0,0.843861,0.376754,0.20,0.51,0.84,1.17,1.50
Electrical_Conductivity,2857.0,1.523588,0.849160,0.10,0.78,1.51,2.28,3.00
Nitrogen_Level,2857.0,88.186209,40.700308,20.00,52.00,88.00,123.00,159.00
Phosphorus_Level,2857.0,48.820441,23.322163,10.00,29.00,49.00,69.00,89.00
Potassium_Level,2857.0,64.694085,32.158062,10.00,37.00,65.00,93.00,119.00
Temperature,2857.0,25.167336,8.621460,10.00,17.61,25.48,32.62,40.00
Humidity,2857.0,59.912226,17.440178,30.03,44.58,60.20,75.71,89.88
Rainfall,2857.0,1596.013297,809.380258,200.36,895.53,1606.49,2288.98,2999.93


In [12]:
categorical_features = [
    "Soil_Type",
    "Crop_Type",
    "Crop_Growth_Stage",
    "Season",
    "Irrigation_Type",
    "Previous_Crop",
    "Region",
    "Recommended_Fertilizer"
]

for col in categorical_features:
    print("\n==========================")
    print(col)
    print("==========================")
    print(fertilizer_df[col].value_counts())


Soil_Type
Soil_Type
Silt     736
Sandy    733
Clay     699
Loamy    689
Name: count, dtype: int64

Crop_Type
Crop_Type
Rice         1429
Sugarcane    1428
Name: count, dtype: int64

Crop_Growth_Stage
Crop_Growth_Stage
Harvest       740
Vegetative    729
Sowing        706
Flowering     682
Name: count, dtype: int64

Season
Season
Zaid      967
Rabi      953
Kharif    937
Name: count, dtype: int64

Irrigation_Type
Irrigation_Type
Drip         736
Sprinkler    717
Rainfed      708
Canal        696
Name: count, dtype: int64

Previous_Crop
Previous_Crop
Wheat        428
Cotton       422
Maize        418
Sugarcane    404
Rice         403
Potato       392
Tomato       390
Name: count, dtype: int64

Region
Region
East       596
North      584
South      575
Central    552
West       550
Name: count, dtype: int64

Recommended_Fertilizer
Recommended_Fertilizer
Urea             912
DAP              838
MOP              394
Compost          281
Zinc Sulphate    203
NPK              182
SSP       

In [13]:
nutrient_by_fertilizer = (
    fertilizer_df
    .groupby("Recommended_Fertilizer")[
        [
            "Nitrogen_Level",
            "Phosphorus_Level",
            "Potassium_Level"
        ]
    ]
    .mean()
    .round(2)
)

display(nutrient_by_fertilizer)

,Nitrogen_Level,Phosphorus_Level,Potassium_Level
Recommended_Fertilizer,,,
Compost,110.51,64.75,78.34
DAP,108.66,26.63,67.56
MOP,111.16,63.90,32.45
NPK,108.12,63.56,79.47
SSP,111.77,65.83,81.17
Urea,42.40,50.20,64.23
Zinc Sulphate,110.55,65.74,81.59


In [14]:
crop_nutrient_analysis = (
    fertilizer_df
    .groupby(
        ["Crop_Type", "Recommended_Fertilizer"]
    )[
        [
            "Nitrogen_Level",
            "Phosphorus_Level",
            "Potassium_Level",
            "Soil_pH",
            "Soil_Moisture"
        ]
    ]
    .mean()
    .round(2)
)

display(crop_nutrient_analysis)

Nitrogen_Level  Phosphorus_Level  \
Crop_Type Recommended_Fertilizer                                     
Rice      Compost                         110.48             63.93   
          DAP                             108.73             26.73   
          MOP                             112.59             63.93   
          NPK                             106.17             63.65   
          SSP                             110.18             70.41   
          Urea                             43.22             49.61   
          Zinc Sulphate                   109.07             65.96   
Sugarcane Compost                         110.54             65.63   
          DAP                             108.60             26.53   
          MOP                             109.79             63.88   
          NPK                             110.66             63.44   
          SSP                             112.67             63.23   
          Urea                             41.56             50.79   
          Zinc Sulphate                   111.95             65.53   

                                  Potassium_Level  Soil_pH  Soil_Moisture  
Crop_Type Recommended_Fertilizer                                           
Rice      Compost                           75.31     5.36          34.32  
          DAP                               69.65     6.54          35.12  
          MOP                               32.16     6.64          33.80  
          NPK                               80.17     6.73          38.10  
          SSP                               82.18     6.67          33.74  
          Urea                              65.37     6.53          35.28  
          Zinc Sulphate                     79.52     7.64          33.33  
Sugarcane Compost                           81.57     5.33          34.33  
          DAP                               65.51     6.56          35.90  
          MOP                               32.72     6.56          37.08  
          NPK                               78.54     6.56          36.63  
          SSP                               80.60     6.57          34.66  
          Urea                              63.08     6.47          35.34  
          Zinc Sulphate                     83.56     7.68          35.21

In [15]:
print(
    fertilizer_df.groupby(
        "Recommended_Fertilizer"
    )[
        [
            "Nitrogen_Level",
            "Phosphorus_Level",
            "Potassium_Level",
            "Fertilizer_Used_Last_Season",
            "Yield_Last_Season"
        ]
    ].agg(["mean", "min", "max"]).round(2)
)

                       Nitrogen_Level          Phosphorus_Level          \
                                 mean min  max             mean min max   
Recommended_Fertilizer                                                    
Compost                        110.51  60  159            64.75  40  89   
DAP                            108.66  60  159            26.63  10  86   
MOP                            111.16  60  159            63.90  40  89   
NPK                            108.12  60  159            63.56  40  89   
SSP                            111.77  62  158            65.83  40  89   
Urea                            42.40  20  154            50.20  10  89   
Zinc Sulphate                  110.55  60  159            65.74  41  89   

                       Potassium_Level          Fertilizer_Used_Last_Season  \
                                  mean min  max                        mean   
Recommended_Fertilizer                                                        
Compost     

In [16]:
target_range_analysis = (
    fertilizer_df
    .groupby("Recommended_Fertilizer")[
        [
            "Nitrogen_Level",
            "Phosphorus_Level",
            "Potassium_Level"
        ]
    ]
    .agg(["min", "max", "mean", "median"])
    .round(2)
)

display(target_range_analysis)

Nitrogen_Level                     Phosphorus_Level  \
                                  min  max    mean median              min   
Recommended_Fertilizer                                                       
Compost                            60  159  110.51  109.0               40   
DAP                                60  159  108.66  109.0               10   
MOP                                60  159  111.16  112.0               40   
NPK                                60  159  108.12  110.5               40   
SSP                                62  158  111.77  112.0               40   
Urea                               20  154   42.40   40.0               10   
Zinc Sulphate                      60  159  110.55  112.0               41   

                                         Potassium_Level                     
                       max   mean median             min  max   mean median  
Recommended_Fertilizer                                                       
Compost                 89  64.75   64.0              40  119  78.34   77.0  
DAP                     86  26.63   25.0              10  119  67.56   69.0  
MOP                     89  63.90   62.0              10  117  32.45   27.0  
NPK                     89  63.56   63.5              40  119  79.47   77.5  
SSP                     89  65.83   65.0              45  115  81.17   81.0  
Urea                    89  50.20   51.0              10  119  64.23   63.0  
Zinc Sulphate           89  65.74   67.0              40  119  81.59   84.0

In [17]:
other_target_analysis = (
    fertilizer_df
    .groupby("Recommended_Fertilizer")[
        [
            "Soil_pH",
            "Soil_Moisture",
            "Organic_Carbon",
            "Electrical_Conductivity",
            "Temperature",
            "Humidity",
            "Rainfall",
            "Fertilizer_Used_Last_Season",
            "Yield_Last_Season"
        ]
    ]
    .mean()
    .round(2)
)

display(other_target_analysis)

,Soil_pH,Soil_Moisture,Organic_Carbon,Electrical_Conductivity,Temperature,Humidity,Rainfall,Fertilizer_Used_Last_Season,Yield_Last_Season
Recommended_Fertilizer,,,,,,,,,
Compost,5.34,34.32,0.87,1.56,25.15,59.05,1598.53,174.49,4.55
DAP,6.55,35.51,0.83,1.52,25.49,60.00,1603.75,174.64,4.42
MOP,6.60,35.48,0.84,1.54,24.58,60.95,1566.74,173.27,4.60
NPK,6.66,37.46,0.84,1.55,24.85,59.86,1633.96,173.03,4.39
SSP,6.61,34.33,0.86,1.42,26.94,57.41,1617.95,179.03,4.58
Urea,6.50,35.31,0.85,1.50,24.98,59.92,1605.18,177.61,4.50
Zinc Sulphate,7.66,34.29,0.84,1.56,25.74,59.32,1537.16,171.92,4.35


In [18]:
crop_percentages = pd.crosstab(
    fertilizer_df["Crop_Type"],
    fertilizer_df["Recommended_Fertilizer"],
    normalize="index"
) * 100

display(crop_percentages.round(2))

Recommended_Fertilizer,Compost,DAP,MOP,NPK,SSP,Urea,Zinc Sulphate
Crop_Type,,,,,,,
Rice,10.15,28.97,13.51,7.21,1.19,32.05,6.93
Sugarcane,9.52,29.69,14.08,5.53,2.10,31.79,7.28


In [19]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, export_text
from sklearn.metrics import accuracy_score, classification_report

X_diag = fertilizer_df[
    [
        "Nitrogen_Level",
        "Phosphorus_Level",
        "Potassium_Level"
    ]
]

y_diag = fertilizer_df["Recommended_Fertilizer"]

X_train_diag, X_test_diag, y_train_diag, y_test_diag = train_test_split(
    X_diag,
    y_diag,
    test_size=0.20,
    random_state=42,
    stratify=y_diag
)

diag_tree = DecisionTreeClassifier(
    max_depth=4,
    random_state=42,
    class_weight="balanced"
)

diag_tree.fit(X_train_diag, y_train_diag)

diag_pred = diag_tree.predict(X_test_diag)

print("NPK-only diagnostic accuracy:")
print(round(accuracy_score(y_test_diag, diag_pred), 4))

print("\nClassification Report:")
print(classification_report(
    y_test_diag,
    diag_pred,
    zero_division=0
))

print("\nLearned Rules:")
print(
    export_text(
        diag_tree,
        feature_names=list(X_diag.columns)
    )
)

NPK-only diagnostic accuracy:
0.722

Classification Report:
               precision    recall  f1-score   support

      Compost       0.09      0.02      0.03        56
          DAP       1.00      0.95      0.98       168
          MOP       1.00      0.87      0.93        79
          NPK       0.00      0.00      0.00        36
          SSP       0.06      1.00      0.11         9
         Urea       1.00      0.95      0.97       183
Zinc Sulphate       0.00      0.00      0.00        41

     accuracy                           0.72       572
    macro avg       0.45      0.54      0.43       572
 weighted avg       0.76      0.72      0.73       572


Learned Rules:
|--- Nitrogen_Level <= 59.50
|   |--- class: Urea
|--- Nitrogen_Level >  59.50
|   |--- Phosphorus_Level <= 39.50
|   |   |--- class: DAP
|   |--- Phosphorus_Level >  39.50
|   |   |--- Potassium_Level <= 39.50
|   |   |   |--- class: MOP
|   |   |--- Potassium_Level >  39.50
|   |   |   |--- Potassium_Level <= 44.

In [20]:
ph_range_analysis = (
    fertilizer_df
    .groupby("Recommended_Fertilizer")["Soil_pH"]
    .agg(["min", "max", "mean", "median"])
    .round(2)
)

display(ph_range_analysis)

,min,max,mean,median
Recommended_Fertilizer,,,,
Compost,4.51,7.75,5.34,5.19
DAP,4.51,8.50,6.55,6.58
MOP,4.51,8.49,6.60,6.62
NPK,5.53,7.79,6.66,6.62
SSP,5.65,7.74,6.61,6.54
Urea,4.50,8.49,6.50,6.51
Zinc Sulphate,5.56,8.49,7.66,7.99


In [21]:
diagnostic_features = [
    "Soil_pH",
    "Soil_Moisture",
    "Organic_Carbon",
    "Electrical_Conductivity",
    "Nitrogen_Level",
    "Phosphorus_Level",
    "Potassium_Level",
    "Temperature",
    "Humidity",
    "Rainfall",
    "Fertilizer_Used_Last_Season",
    "Yield_Last_Season"
]

full_range_analysis = (
    fertilizer_df
    .groupby("Recommended_Fertilizer")[diagnostic_features]
    .agg(["min", "max", "mean"])
    .round(2)
)

display(full_range_analysis)

Soil_pH             Soil_Moisture                \
                           min   max  mean           min    max   mean   
Recommended_Fertilizer                                                   
Compost                   4.51  7.75  5.34         10.03  59.94  34.32   
DAP                       4.51  8.50  6.55         10.02  59.86  35.51   
MOP                       4.51  8.49  6.60         10.03  60.00  35.48   
NPK                       5.53  7.79  6.66         10.07  59.83  37.46   
SSP                       5.65  7.74  6.61         10.22  59.54  34.33   
Urea                      4.50  8.49  6.50         10.00  60.00  35.31   
Zinc Sulphate             5.56  8.49  7.66         10.08  59.80  34.29   

                       Organic_Carbon             Electrical_Conductivity  \
                                  min   max  mean                     min   
Recommended_Fertilizer                                                      
Compost                          0.20  1.50  0.87                    0.11   
DAP                              0.20  1.50  0.83                    0.10   
MOP                              0.21  1.50  0.84                    0.11   
NPK                              0.21  1.49  0.84                    0.11   
SSP                              0.23  1.42  0.86                    0.10   
Urea                             0.20  1.50  0.85                    0.10   
Zinc Sulphate                    0.21  1.50  0.84                    0.11   

                        ... Humidity Rainfall                    \
                        ...     mean      min      max     mean   
Recommended_Fertilizer  ...                                       
Compost                 ...    59.05   200.36  2978.61  1598.53   
DAP                     ...    60.00   201.34  2999.93  1603.75   
MOP                     ...    60.95   206.19  2997.23  1566.74   
NPK                     ...    59.86   239.03  2991.13  1633.96   
SSP                     ...    57.41   257.25  2960.48  1617.95   
Urea                    ...    59.92   201.52  2996.15  1605.18   
Zinc Sulphate           ...    59.32   236.56  2978.79  1537.16   

                       Fertilizer_Used_Last_Season                  \
                                               min     max    mean   
Recommended_Fertilizer                                               
Compost                                      50.37  298.17  174.49   
DAP                                          50.01  300.00  174.64   
MOP                                          50.86  299.63  173.27   
NPK                                          51.56  299.24  173.03   
SSP                                          56.26  294.83  179.03   
Urea                                         50.08  299.71  177.61   
Zinc Sulphate                                52.09  299.71  171.92   

                       Yield_Last_Season              
                                     min   max  mean  
Recommended_Fertilizer                                
Compost                             1.02  7.99  4.55  
DAP                                 1.00  8.00  4.42  
MOP                                 1.01  7.98  4.60  
NPK                                 1.02  7.94  4.39  
SSP                                 1.05  7.99  4.58  
Urea                                1.00  7.99  4.50  
Zinc Sulphate                       1.07  8.00  4.35  

[7 rows x 36 columns]

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, export_text
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    classification_report
)

X_diag2 = fertilizer_df[
    [
        "Nitrogen_Level",
        "Phosphorus_Level",
        "Potassium_Level",
        "Soil_pH",
        "Electrical_Conductivity",
        "Organic_Carbon",
        "Soil_Moisture"
    ]
]

y_diag2 = fertilizer_df["Recommended_Fertilizer"]

X_train2, X_test2, y_train2, y_test2 = train_test_split(
    X_diag2,
    y_diag2,
    test_size=0.20,
    random_state=42,
    stratify=y_diag2
)

diag_tree2 = DecisionTreeClassifier(
    max_depth=7,
    class_weight="balanced",
    random_state=42
)

diag_tree2.fit(X_train2, y_train2)

pred2 = diag_tree2.predict(X_test2)

print("Accuracy:")
print(round(accuracy_score(y_test2, pred2), 4))

print("\nBalanced Accuracy:")
print(round(balanced_accuracy_score(y_test2, pred2), 4))

print("\nClassification Report:")
print(
    classification_report(
        y_test2,
        pred2,
        zero_division=0
    )
)

print("\nLearned Rules:")
print(
    export_text(
        diag_tree2,
        feature_names=list(X_diag2.columns),
        max_depth=7
    )
)

Accuracy:
0.8759

Balanced Accuracy:
0.7334

Classification Report:
               precision    recall  f1-score   support

      Compost       1.00      0.79      0.88        56
          DAP       0.99      0.95      0.97       168
          MOP       1.00      0.87      0.93        79
          NPK       0.41      0.67      0.51        36
          SSP       0.06      0.22      0.09         9
         Urea       1.00      0.95      0.97       183
Zinc Sulphate       0.97      0.68      0.80        41

     accuracy                           0.88       572
    macro avg       0.77      0.73      0.74       572
 weighted avg       0.94      0.88      0.90       572


Learned Rules:
|--- Nitrogen_Level <= 59.50
|   |--- class: Urea
|--- Nitrogen_Level >  59.50
|   |--- Phosphorus_Level <= 39.50
|   |   |--- class: DAP
|   |--- Phosphorus_Level >  39.50
|   |   |--- Potassium_Level <= 39.50
|   |   |   |--- class: MOP
|   |   |--- Potassium_Level >  39.50
|   |   |   |--- Soil_pH <= 5.5

In [23]:
feature_importance = pd.DataFrame({
    "Feature": X_diag2.columns,
    "Importance": diag_tree2.feature_importances_
}).sort_values(
    "Importance",
    ascending=False
)

display(feature_importance)

,Feature,Importance
3,Soil_pH,0.336555
0,Nitrogen_Level,0.229409
1,Phosphorus_Level,0.228167
2,Potassium_Level,0.196594
4,Electrical_Conductivity,0.009275
5,Organic_Carbon,0.000000
6,Soil_Moisture,0.000000


## Dataset Analysis Conclusion

- Original dataset size: 10,000 records.
- Client-specific dataset contains 2,857 records:
  - Rice: 1,429
  - Sugarcane: 1,428
- Target variable: Recommended_Fertilizer.
- Seven fertilizer classes are available:
  Urea, DAP, MOP, Compost, NPK, SSP, Zinc Sulphate.
- The target is imbalanced, especially for SSP.
- Diagnostic Decision Tree analysis showed that the fertilizer labels are
  strongly determined by Soil pH, Nitrogen, Phosphorus and Potassium.
- Soil Moisture and Organic Carbon provided almost no predictive contribution
  in the diagnostic tree.
- Rice and Sugarcane show very similar fertilizer-class distributions.
- The dataset therefore appears to follow structured/synthetic fertilizer
  recommendation rules rather than representing fully validated crop-specific
  field observations.
- The dataset will be used for prototype ML experimentation and decision-support
  development only.
- Real agronomic validation is required before operational fertilizer advice.